### PR TITLE
Fix SSE-C source decryption handling

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Fix SSE-C source decryption handling

## Motivation and Context
Without this fix, we have room for two different type of errors.

- The source is encrypted and we didn't provide any source encryption keys

This results in Incomplete body error to be returned back to the client
since the source is encrypted and we gave the reader as is to the object
layer which was of a decrypted value leading to "IncompleteBody"

- The source is not encrypted and we provided source encryption keys.

This results in a corrupted object on the destination which is
considered encrypted but cannot be read by the server and returns
the following error.

```
<Error><Code>XMinioObjectTampered</Code><Message>The requested object
was modified and may be compromised</Message><Resource>/id-platform-gamma/
</Resource><RequestId>155EDC3E86BFD4DA</RequestId><HostId>3L137</HostId>
</Error>
```

## Regression
Maybe

## How Has This Been Tested?
Using the following example code. To start with let's create an encrypted object and an unencrypted object

```
~ mc cp --encrypt-key="myminio/=32byteslongsecretkeymustbegiven2" /etc/issue myminio/id-platform-gamma/issue-enc
~ mc cp /etc/issue myminio/id-platform-gamma/issue-noenc
```

Use the following code on the master now if you can see since we are not passing CopySSE keys for an encrypted object, it should throw an error which it does but the error is weird and returns "IncompleteBody" and we can see log messages regarding this. This is the first variation of the bug. There is no destination object that is created.  With this PR a proper error is returned requesting proper encryption parameters.
```go
package main

import (
	"crypto/tls"
	"fmt"
	"net/http"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"
)

func main() {
	creds := credentials.NewStaticCredentials("minio", "minio123", "")
	newSession := session.New()
	http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
	s3Config := &aws.Config{
		Credentials:      creds,
		Endpoint:         aws.String("https://127.0.0.1:9012"),
		Region:           aws.String("us-east-1"),
		S3ForcePathStyle: aws.Bool(true),
		LogLevel:         aws.LogLevel(aws.LogDebug),
	}
	// Create an S3 service object in the default region.
	s3Client := s3.New(newSession, s3Config)
	params := &s3.CopyObjectInput{
		SSECustomerAlgorithm:           aws.String("AES256"),                           // header x-amz-server-side-encryption-customer-algorithm
		SSECustomerKey:                 aws.String("32byteslongsecretkeymustbegiven2"), // header x-amz-server-side-encryption-customer-key
		CopySource:                     aws.String("/id-platform-gamma/issue-enc"),
		Bucket:                         aws.String("id-platform-gamma"),
		Key:                            aws.String("issue1"),
	}
	req, output := s3Client.CopyObjectRequest(params)
	err := req.Send()
	if err != nil {
		fmt.Println(err)
		return
	}
	fmt.Println(output)
}
```

Use the following code on the master now if you can see since we are passing CopySSE keys for an unencrypted object, it should throw an error but it doesn't and we see that destination object gets created nevertheless and we cannot read the object anymore. But with this PR this doesn't happen and we cannot read the object anymore. This is the second variation of the bug. 
```go
package main

import (
	"crypto/tls"
	"fmt"
	"net/http"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"
)

func main() {
	creds := credentials.NewStaticCredentials("minio", "minio123", "")
	newSession := session.New()
	http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
	s3Config := &aws.Config{
		Credentials:      creds,
		Endpoint:         aws.String("https://127.0.0.1:9012"),
		Region:           aws.String("us-east-1"),
		S3ForcePathStyle: aws.Bool(true),
		LogLevel:         aws.LogLevel(aws.LogDebug),
	}
	// Create an S3 service object in the default region.
	s3Client := s3.New(newSession, s3Config)
	params := &s3.CopyObjectInput{
		CopySourceSSECustomerKey:       aws.String("32byteslongsecretkeymustbegiven2"),
		CopySourceSSECustomerAlgorithm: aws.String("AES256"),
		SSECustomerAlgorithm:           aws.String("AES256"),                           // header x-amz-server-side-encryption-customer-algorithm
		SSECustomerKey:                 aws.String("32byteslongsecretkeymustbegiven2"), // header x-amz-server-side-encryption-customer-key
		CopySource:                     aws.String("/id-platform-gamma/issue-noenc"),
		Bucket:                         aws.String("id-platform-gamma"),
		Key:                            aws.String("issue1"),
	}
	req, output := s3Client.CopyObjectRequest(params)
	err := req.Send()
	if err != nil {
		fmt.Println(err)
		return
	}
	fmt.Println(output)
}
```

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/object-handlers.go [hunk 1]
```diff
@@ -808,9 +808,21 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
 	var encMetadata = make(map[string]string)
 	if objectAPI.IsEncryptionSupported() && !srcInfo.IsCompressed() {
+		// Encryption parameters not applicable for this object.
+		if !crypto.IsEncrypted(srcInfo.UserDefined) && crypto.SSECopy.IsRequested(r.Header) {
+			writeErrorResponse(w, toAPIErrorCode(errInvalidEncryptionParameters), r.URL)
+			return
+		}
+
+		// Encryption parameters not present for this object.
+		if crypto.IsEncrypted(srcInfo.UserDefined) && !crypto.SSECopy.IsRequested(r.Header) {
```
Comment (poornas) [inline cmd/object-handlers.go]: @harshavardhana , you will have to check if crypto.SSEC.IsEncrypted(srcInfo.UserDefined) as SSECopy headers not required if it source is encrypted with SSE-S3

Comment (harshavardhana) [reply cmd/object-handlers.go]: okay cool

**Comment by harshavardhana** [comment]
Similarly, with a little bigger file lets say 3mb, we can reproduce this

```
$ mc cat --encrypt-key="myminio-local1-ssl/=32byteslongsecretkeymustbegiven2" myminio-local1-ssl/id-platform-gamma/issue1 > /dev/null
mc: <ERROR> Unable to read from `myminio-local1-ssl/id-platform-gamma/issue1`. Input reader closed pre-maturely. Expected `3144192` bytes, but only received `3080192` bytes.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

File: cmd/object-handlers.go [hunk 2]
```diff
@@ -808,9 +808,21 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+		// Encryption parameters not present for this object.
+		if crypto.SSEC.IsEncrypted(srcInfo.UserDefined) && !crypto.SSECopy.IsRequested(r.Header) {
+			writeErrorResponse(w, ErrInvalidSSECustomerAlgorithm, r.URL)
+			return
+		}
+
 		var oldKey, newKey []byte
 		sseCopyS3 := crypto.S3.IsEncrypted(srcInfo.UserDefined)
-		sseCopyC := crypto.SSECopy.IsRequested(r.Header)
+		sseCopyC := crypto.SSEC.IsEncrypted(srcInfo.UserDefined) && crypto.SSECopy.IsRequested(r.Header)
```
Comment (kannappanr) [inline cmd/object-handlers.go:825]: `sseCopyC := crypto.SSEC.IsEncrypted(srcInfo.UserDefined)` is sufficient. At this point, if `crypto.SSEC.IsEncrypted(srcInfo.UserDefined)` is true then `crypto.SSECopy.IsRequested(r.Header)` will also be true.

Comment (harshavardhana) [reply cmd/object-handlers.go:825]: This to be more explicit about the process. It is a complex piece of code I don't want any assumptions.

**Comment by aead** [comment]
@harshavardhana Can you re-run CI? Failures again probably not related...

**Review by aead — APPROVED** [review]

**Review by vadmeste — APPROVED** [review]
LGTM, minio-py/minio-go/aws-sdk-java with TLS tests were ran against it

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Pull request merged**